### PR TITLE
Control+C handling

### PR DESCRIPTION
**Pull request by aalekseyev**
This PR deals with problematic behavior of dune when it's interrupted with Ctrl+C while running a build in polling mode.
Before this PR, dune would just fail with "assert false" and keep watching, among other weird behaviors.

After this PR, killing with Ctrl+C should just terminate dune quickly.

### CODE REVIEW

**Comment by aalekseyev** [comment]
@diml, this should be ready for review now. The first two commits implement the actual fix, and the rest are a cleanup with no significant intended change of behavior.

File: src/scheduler.ml [hunk 1]
```diff
@@ -675,64 +728,38 @@ let poll ?log ?config ~once ~finally () =
...
+    match res with
+    | Ok (Ok ()) ->
+      wait "Success" |> after_wait
+    | Ok (Error Got_signal) ->
+      (Already_reported, None)
+    | Ok (Error Never) ->
+      wait "Had errors" |> after_wait
+    | Ok (Error Files_changed) ->
+      loop ()
+    | Error exn -> (exn, Some (Printexc.get_raw_backtrace ()))
```
Comment (diml) [inline src/scheduler.ml]: We should probably capture this backtrace earlier, i.e. when constructing the original `Error exn`

File: src/scheduler.ml [hunk 2]
```diff
@@ -560,14 +559,36 @@ let got_signal t signal =
...
+  while Event.pending_jobs () > 0 do
+    match Event.next () with
+    | Signal signal ->
+      got_signal t signal;
+      saw_signal := Got_signal;
+    | _ -> ()
+  done;
+  !saw_signal
+
 let go_rec t =
```
Comment (diml) [inline src/scheduler.ml]: We can probably get rid of this layer

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by diml** [comment]
Thanks, the result does look easier to follow. BTW, looking again at this file, the function names are not great: we have `go_rec` that is called by `run`, itself called by `run_and_cleanup` which is then called by both `go` and `poll`. I suggest to put the first three functions into a sub-module `Run_once` or `Main_loop` that only exposes `run_and_cleanup` and maybe rename `go_rec` to something else, such as `main_loop`.

**Comment by aalekseyev** [comment]
Pushed the changes.
I renamed `go_rec` to `pump_events` because that seems like what it's doing.

File: src/scheduler.ml [hunk 3]
```diff
@@ -628,45 +615,127 @@ let prepare ?(log=Log.no_log) ?(config=Config.default) () =
...
+          Fiber.Ivar.fill job.ivar status
+          >>= fun () ->
+          restart_waiting_for_available_job t
+          >>= fun () ->
+          pump_events t
+        | Files_changed ->
+          Fiber.return Files_changed
+        | Signal signal ->
+          got_signal t signal;
+          Fiber.return (Got_signal : pump_events_result)
```
Comment (diml) [inline src/scheduler.ml]: We should be able to get rid of this type annotation now:

```suggestion
          Fiber.return Got_signal
```

**Review by ghost — APPROVED** [review]

**Pull request merged**